File: src/main/java/fr/maxlego08/satisfactorydle/command/EmbedHelper.java
package fr.maxlego08.satisfactorydle.command;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import fr.maxlego08.satisfactorydle.Messages;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;

import java.awt.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

public final class EmbedHelper {

    public static final Color COLOR_SUCCESS = new Color(0x57F287);
    public static final Color COLOR_ERROR = new Color(0xED4245);
    public static final Color COLOR_WARNING = new Color(0xFEE75C);
    public static final Color COLOR_INFO = new Color(0x5865F2);

    private static final String FOOTER_ICON = "https://satisfactorydle.net/favicon-96x96.png";
    private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("dd/MM/yyyy");

    private static final Map<String, String> COLUMN_TO_FIELD_KEY = Map.of(
            "power_consumption", "power",
            "is_alternate", "alternate",
            "unlocked_items_count", "unlocked_items",
            "used_in_count", "used_in_count"
    );

    private EmbedHelper() {
    }

    public static void applyFooter(EmbedBuilder embed, String text) {
        String date = LocalDate.now().format(DATE_FORMAT);
        String footer = text != null && !text.isEmpty() ? text + " | satisfactorydle.net | " + date : "satisfactorydle.net | " + date;
        embed.setFooter(footer, FOOTER_ICON);
    }

    public static void replyError(SlashCommandInteractionEvent event, Messages messages, String message) {
        EmbedBuilder embed = new EmbedBuilder().setColor(COLOR_ERROR).setTitle(messages.get("error.title")).setDescription(message);
        applyFooter(embed, null);
        event.getHook().editOriginalEmbeds(embed.build()).queue();
    }

    public static void addAnswerFields(EmbedBuilder embed, JsonObject answer, String mode, Messages messages) {
        switch (mode) {
            case "classic" -> {
                addFieldIfPresent(embed, messages.get("field.category"), answer, "category", true);
                addFieldIfPresent(embed, messages.get("field.tier"), answer, "tier", true);
                addFieldIfPresent(embed, messages.get("field.stack_size"), answer, "stack_size", true);
                addFieldIfPresent(embed, messages.get("field.form"), answer, "form", true);
                addFieldIfPresent(embed, messages.get("field.sink_points"), answer, "sink_points", true);
                addFieldIfPresent(embed, messages.get("field.used_in_count"), answer, "used_in_count", true);
            }
            case "item" -> {
                addFieldIfPresent(embed, messages.get("field.category"), answer, "category", true);
                addFieldIfPresent(embed, messages.get("field.tier"), answer, "tier", true);
                addFieldIfPresent(embed, messages.get("field.form"), answer, "form", true);
                addFieldIfPresent(embed, messages.get("field.stack_size"), answer, "stack_size", true);
                addFieldIfPresent(embed, messages.get("field.sink_points"), answer, "sink_points", true);
            }
            case "building" -> {
                addFieldIfPresent(embed, messages.get("field.category"), answer, "category", true);
                addFieldIfPresent(embed, messages.get("field.tier"), answer, "tier", true);
                addFieldIfPresent(embed, messages.get("field.power"), answer, "power_consumption", true);
            }
            case "recipe" -> {
                addFieldIfPresent(embed, messages.get("field.building"), answer, "building", true);
                if (answer.has("is_alternate") && !answer.get("is_alternate").isJsonNull()) {
                    embed.addField(messages.get("field.alternate"), answer.get("is_alternate").getAsBoolean() ? messages.get("common.yes") : messages.get("common.no"), true);
                }
                addInputItems(embed, answer, messages);
            }
            case "creature" -> {
                addFieldIfPresent(embed, messages.get("field.hostility"), answer, "hostility", true);
                addFieldIfPresent(embed, messages.get("field.biome"), answer, "biome", true);
                addFieldIfPresent(embed, messages.get("field.type"), answer, "type", true);
            }
            case "milestone" -> {
                addFieldIfPresent(embed, messages.get("field.source"), answer, "source", true);
                addFieldIfPresent(embed, messages.get("field.tier"), answer, "tier", true);
                addFieldIfPresent(embed, messages.get("field.unlocked_items"), answer, "unlocked_items_count", true);
            }
        }

        if (hasValue(answer, "description")) {
            String desc = answer.get("description").getAsString();
            if (desc.length() > 200) desc = desc.substring(0, 200) + "...";
            embed.addField(messages.get("field.description"), desc, false);
        }
    }

    public static void addHintFields(EmbedBuilder embed, JsonObject hints, Messages messages) {
        StringBuilder unlocked = new StringBuilder();
        StringBuilder locked = new StringBuilder();

        for (String key : hints.keySet()) {
            if (key.equals("locked") || key.equals("image_url")) continue;
            JsonElement value = hints.get(key);
            if (value == null || value.isJsonNull()) continue;

            String label = formatLabel(key);

            if (key.equals("reveal_image")) {
                if (value.getAsBoolean() && hasValue(hints, "image_url")) {
                    embed.setImage(hints.get("image_url").getAsString());
                    unlocked.append("**Image:** ").append(messages.get("hints.image_revealed")).append("\n");
                }
            } else if (key.equals("is_alternate")) {
                unlocked.append("**").append(label).append(":** ").append(value.getAsBoolean() ? messages.get("common.yes") : messages.get("common.no")).append("\n");
            } else if (value.isJsonArray()) {
                unlocked.append("**").append(label).append(":** ").append(formatJsonArray(value.getAsJsonArray())).append("\n");
            } else {
                unlocked.append("**").append(label).append(":** ").append(jsonToString(value, messages)).append("\n");
            }
        }

        if (hints.has("locked") && hints.get("locked").isJsonArray()) {
            for (JsonElement item : hints.getAsJsonArray("locked")) {
                JsonObject lock = item.getAsJsonObject();
                String label = lock.get("label").getAsString();
                int remaining = lock.get("remaining").getAsInt();
                String guessWord = remaining > 1 ? messages.get("common.guesses") : messages.get("common.guess");
                locked.append("**").append(label).append("** - ").append(messages.get("hints.unlock_in", "remaining", remaining, "guess_word", guessWord)).append("\n");
            }
        }

        if (!unlocked.isEmpty()) {
            embed.addField(messages.get("hints.title"), unlocked.toString(), false);
        }
        if (!locked.isEmpty()) {
            embed.addField(messages.get("hints.locked_title"), locked.toString(), false);
        }
    }

    public static void addInputItems(EmbedBuilder embed, JsonObject json, Messages messages) {
        if (!json.has("input_items") || !json.get("input_items").isJsonArray()) return;
        JsonArray inputs = json.getAsJsonArray("input_items");
        if (inputs.isEmpty()) return;

        StringBuilder sb = new StringBuilder();
        for (JsonElement element : inputs) {
            JsonObject item = element.getAsJsonObject();
            sb.append("- ").append(item.get("slug").getAsString());
            if (item.has("amount")) sb.append(" x").append(item.get("amount").getAsInt());
            sb.append("\n");
        }
        embed.addField(messages.get("field.input_items"), sb.toString(), false);
    }

    public static void addFieldIfPresent(EmbedBuilder embed, String label, JsonObject json, String key, boolean inline) {
        if (hasValue(json, key)) {
            embed.addField(label, jsonToStringRaw(json.get(key)), inline);
        }
    }

    public static boolean hasValue(JsonObject json, String key) {
        return json.has(key) && !json.get(key).isJsonNull();
    }

    public static String jsonToString(JsonElement element, Messages messages) {
        if (element.isJsonPrimitive()) {
            JsonPrimitive prim = element.getAsJsonPrimitive();
            if (prim.isBoolean()) return prim.getAsBoolean() ? messages.get("common.yes") : messages.get("common.no");
            if (prim.isNumber()) return prim.getAsNumber().toString();
            return prim.getAsString();
        }
        return element.toString();
    }

    public static String jsonToStringRaw(JsonElement element) {
        if (element.isJsonPrimitive()) {
            JsonPrimitive prim = element.getAsJsonPrimitive();
            if (prim.isBoolean()) return prim.getAsBoolean() ? "Yes" : "No";
            if (prim.isNumber()) return prim.getAsNumber().toString();
            return prim.getAsString();
        }
        return element.toString();
    }

    public static String formatJsonArray(JsonArray array) {
        StringBuilder sb = new StringBuilder();
        for (JsonElement element : array) {
            if (!sb.isEmpty()) sb.append(", ");
            if (element.isJsonObject()) {
                JsonObject obj = element.getAsJsonObject();
                sb.append(obj.get("slug").getAsString());
                if (obj.has("amount")) sb.append(" x").append(obj.get("amount").getAsInt());
            } else {
                sb.append(jsonToStringRaw(element));
            }
        }
        return sb.toString();
    }

    public static String formatLabel(String key) {
        String result = key.replace("_", " ");
        return result.substring(0, 1).toUpperCase() + result.substring(1);
    }

    public static void addGuessHistory(EmbedBuilder embed, JsonArray guesses, String mode, Messages messages) {
        if (guesses.isEmpty()) return;

        List<String> columns = getComparisonColumns(mode);
        if (columns.isEmpty()) return;

        // Column headers in italics
        StringBuilder header = new StringBuilder("*");
        for (int i = 0; i < columns.size(); i++) {
            if (i > 0) header.append(" | ");
            String fieldKey = COLUMN_TO_FIELD_KEY.getOrDefault(columns.get(i), columns.get(i));
            header.append(messages.get("field." + fieldKey));
        }
        header.append("*\n");

        StringBuilder sb = new StringBuilder(header);
        int start = Math.max(0, guesses.size() - 10);

        for (int i = start; i < guesses.size(); i++) {
            JsonObject guess = guesses.get(i).getAsJsonObject();
            String name = guess.get("name").getAsString();
            boolean correct = guess.get("correct").getAsBoolean();

            sb.append("\n").append(correct ? "\u2705" : "\u274C").append(" **").append(name).append("**\n");

            if (guess.has("comparisons") && !guess.get("comparisons").isJsonNull()) {
                JsonObject comparisons = guess.getAsJsonObject("comparisons");
                StringBuilder row = new StringBuilder();
                for (String col : columns) {
                    if (!row.isEmpty()) row.append(" | ");
                    if (comparisons.has(col) && !comparisons.get(col).isJsonNull()) {
                        JsonObject comp = comparisons.getAsJsonObject(col);
                        String result = comp.has("result") ? comp.get("result").getAsString() : "wrong";
                        String value = comp.has("value") && !comp.get("value").isJsonNull()
                                ? comp.get("value").getAsString() : "?";
                        row.append(resultEmoji(result)).append(" ").append(value);
                    } else {
                        row.append("\u2B1C ?");
                    }
                }
                sb.append(row).append("\n");
            }
        }

        String title = messages.get("guess.history_title");
        if (start > 0) {
            title += " (" + messages.get("guess.history_last", "count", 10) + ")";
        }

        String value = sb.toString();
        if (value.length() > 1024) {
            value = value.substring(0, 1020) + "...";
        }

        embed.addField(title, value, false);
    }

    private static List<String> getComparisonColumns(String mode) {
        return switch (mode) {
            case "classic" -> List.of("category", "tier", "stack_size", "form", "sink_points", "used_in_count");
            case "item" -> List.of("category", "tier", "form", "stack_size", "sink_points");
            case "building" -> List.of("category", "tier", "power_consumption");
            case "recipe" -> List.of("building", "is_alternate", "tier");
            case "creature" -> List.of("hostility", "biome", "type");
            case "milestone" -> List.of("source", "tier", "unlocked_items_count");
            default -> List.of();
        };
    }

    private static String resultEmoji(String result) {
        return switch (result) {
            case "correct" -> "\uD83D\uDFE9";
            case "higher" -> "\u2B06\uFE0F";
            case "lower" -> "\u2B07\uFE0F";
            case "partial" -> "\uD83D\uDFE7";
            default -> "\uD83D\uDFE5";
        };
    }
}

File: app/commands/embed_helper.py
from __future__ import annotations

from datetime import date
from typing import Any

import discord


COLOR_SUCCESS = discord.Color(0x57F287)
COLOR_ERROR = discord.Color(0xED4245)
COLOR_WARNING = discord.Color(0xFEE75C)
COLOR_INFO = discord.Color(0x5865F2)

FOOTER_ICON = "https://satisfactorydle.net/favicon-96x96.png"

COLUMN_TO_FIELD_KEY = {
    "power_consumption": "power",
    "is_alternate": "alternate",
    "unlocked_items_count": "unlocked_items",
    "used_in_count": "used_in_count",
}


def apply_footer(embed: discord.Embed, text: str | None) -> None:
    current_date = date.today().strftime("%d/%m/%Y")
    footer = (
        f"{text} | satisfactorydle.net | {current_date}"
        if text
        else f"satisfactorydle.net | {current_date}"
    )
    embed.set_footer(text=footer, icon_url=FOOTER_ICON)


def build_error_embed(messages: Any, message: str) -> discord.Embed:
    embed = discord.Embed(color=COLOR_ERROR, title=messages.get("error.title"), description=message)
    apply_footer(embed, None)
    return embed


def has_value(data: dict[str, Any], key: str) -> bool:
    return key in data and data[key] is not None


def json_to_string(value: Any, messages: Any) -> str:
    if isinstance(value, bool):
        return messages.get("common.yes") if value else messages.get("common.no")
    return str(value)


def json_to_string_raw(value: Any) -> str:
    if isinstance(value, bool):
        return "Yes" if value else "No"
    return str(value)


def format_json_array(values: list[Any]) -> str:
    rendered: list[str] = []
    for value in values:
        if isinstance(value, dict):
            part = value.get("slug", "")
            if "amount" in value:
                part += f" x{value['amount']}"
            rendered.append(part)
        else:
            rendered.append(json_to_string_raw(value))
    return ", ".join(rendered)


def format_label(key: str) -> str:
    label = key.replace("_", " ")
    return label[:1].upper() + label[1:] if label else key


def format_tier_value(value: str) -> str:
    return "MAM" if value == "-1" else value


def result_emoji(result: str) -> str:
    return {
        "correct": "🟩",
        "higher": "⬆️",
        "lower": "⬇️",
        "partial": "🟧",
    }.get(result, "🟥")


def get_comparison_columns(mode: str) -> list[str]:
    mapping = {
        "classic": ["category", "tier", "stack_size", "form", "sink_points", "used_in_count"],
        "item": ["category", "tier", "form", "stack_size", "sink_points"],
        "building": ["category", "tier", "power_consumption"],
        "recipe": ["building", "is_alternate", "tier"],
        "creature": ["hostility", "biome", "type"],
        "milestone": ["source", "tier", "unlocked_items_count"],
    }
    return mapping.get(mode, [])


def add_field_if_present(embed: discord.Embed, label: str, data: dict[str, Any], key: str, inline: bool) -> None:
    if has_value(data, key):
        value = json_to_string_raw(data[key])
        if key == "tier":
            value = format_tier_value(value)
        embed.add_field(name=label, value=value, inline=inline)


def add_input_items(embed: discord.Embed, data: dict[str, Any], messages: Any) -> None:
    input_items = data.get("input_items")
    if not isinstance(input_items, list) or not input_items:
        return

    rows = []
    for item in input_items:
        row = f"- {item.get('slug', '')}"
        if "amount" in item:
            row += f" x{item['amount']}"
        rows.append(row)
    embed.add_field(name=messages.get("field.input_items"), value="\n".join(rows), inline=False)


def add_hint_fields(embed: discord.Embed, hints: dict[str, Any], messages: Any) -> None:
    unlocked_rows: list[str] = []
    locked_rows: list[str] = []

    for key, value in hints.items():
        if key in {"locked", "image_url"} or value is None:
            continue

        field_key = COLUMN_TO_FIELD_KEY.get(key, key)
        label = messages.get(f"field.{field_key}")
        if label == f"field.{field_key}":
            label = format_label(key)

        if key == "reveal_image":
            if bool(value) and has_value(hints, "image_url"):
                embed.set_image(url=str(hints["image_url"]))
                unlocked_rows.append(f"**Image:** {messages.get('hints.image_revealed')}")
        elif key == "is_alternate":
            display = messages.get("common.yes") if bool(value) else messages.get("common.no")
            unlocked_rows.append(f"**{label}:** {display}")
        elif isinstance(value, list):
            unlocked_rows.append(f"**{label}:** {format_json_array(value)}")
        else:
            display = json_to_string(value, messages)
            if key == "tier":
                display = format_tier_value(display)
            unlocked_rows.append(f"**{label}:** {display}")

    locked = hints.get("locked")
    if isinstance(locked, list):
        for item in locked:
            label = item.get("label", "?")
            remaining = int(item.get("remaining", 0))
            guess_word = messages.get("common.guesses") if remaining > 1 else messages.get("common.guess")
            locked_rows.append(f"**{label}** - {messages.get('hints.unlock_in', 'remaining', remaining, 'guess_word', guess_word)}")

    if unlocked_rows:
        embed.add_field(name=messages.get("hints.title"), value="\n".join(unlocked_rows), inline=False)
    if locked_rows:
        embed.add_field(name=messages.get("hints.locked_title"), value="\n".join(locked_rows), inline=False)


def add_answer_fields(embed: discord.Embed, answer: dict[str, Any], mode: str, messages: Any) -> None:
    if mode == "classic":
        add_field_if_present(embed, messages.get("field.category"), answer, "category", True)
        add_field_if_present(embed, messages.get("field.tier"), answer, "tier", True)
        add_field_if_present(embed, messages.get("field.stack_size"), answer, "stack_size", True)
        add_field_if_present(embed, messages.get("field.form"), answer, "form", True)
        add_field_if_present(embed, messages.get("field.sink_points"), answer, "sink_points", True)
        add_field_if_present(embed, messages.get("field.used_in_count"), answer, "used_in_count", True)
    elif mode == "item":
        add_field_if_present(embed, messages.get("field.category"), answer, "category", True)
        add_field_if_present(embed, messages.get("field.tier"), answer, "tier", True)
        add_field_if_present(embed, messages.get("field.form"), answer, "form", True)
        add_field_if_present(embed, messages.get("field.stack_size"), answer, "stack_size", True)
        add_field_if_present(embed, messages.get("field.sink_points"), answer, "sink_points", True)
    elif mode == "building":
        add_field_if_present(embed, messages.get("field.category"), answer, "category", True)
        add_field_if_present(embed, messages.get("field.tier"), answer, "tier", True)
        add_field_if_present(embed, messages.get("field.power"), answer, "power_consumption", True)
    elif mode == "recipe":
        add_field_if_present(embed, messages.get("field.building"), answer, "building", True)
        if has_value(answer, "is_alternate"):
            value = messages.get("common.yes") if bool(answer["is_alternate"]) else messages.get("common.no")
            embed.add_field(name=messages.get("field.alternate"), value=value, inline=True)
        add_input_items(embed, answer, messages)
    elif mode == "creature":
        add_field_if_present(embed, messages.get("field.hostility"), answer, "hostility", True)
        add_field_if_present(embed, messages.get("field.biome"), answer, "biome", True)
        add_field_if_present(embed, messages.get("field.type"), answer, "type", True)
    elif mode == "milestone":
        add_field_if_present(embed, messages.get("field.source"), answer, "source", True)
        add_field_if_present(embed, messages.get("field.tier"), answer, "tier", True)
        add_field_if_present(embed, messages.get("field.unlocked_items"), answer, "unlocked_items_count", True)

    if has_value(answer, "description"):
        desc = str(answer["description"])
        if len(desc) > 200:
            desc = desc[:200] + "..."
        embed.add_field(name=messages.get("field.description"), value=desc, inline=False)


def add_guess_history(embed: discord.Embed, guesses: list[dict[str, Any]], mode: str, messages: Any) -> None:
    if not guesses:
        return

    columns = get_comparison_columns(mode)
    if not columns:
        return

    headers: list[str] = []
    for column in columns:
        field_key = COLUMN_TO_FIELD_KEY.get(column, column)
        headers.append(messages.get(f"field.{field_key}"))

    lines = [f"*{' | '.join(headers)}*"]
    start = max(0, len(guesses) - 10)

    for guess in guesses[start:]:
        correct = bool(guess.get("correct", False))
        name = str(guess.get("name", "?"))
        lines.append(f"\n{'✅' if correct else '❌'} **{name}**")

        comparisons = guess.get("comparisons")
        if isinstance(comparisons, dict):
            row_parts: list[str] = []
            for column in columns:
                comp = comparisons.get(column)
                if isinstance(comp, dict):
                    result = str(comp.get("result", "wrong"))
                    value = str(comp.get("value", "?"))
                    if column == "tier":
                        value = format_tier_value(value)
                    row_parts.append(f"{result_emoji(result)} {value}")
                else:
                    row_parts.append("⬜ ?")
            lines.append(" | ".join(row_parts))

    title = messages.get("guess.history_title")
    if start > 0:
        title += f" ({messages.get('guess.history_last', 'count', 10)})"

    value = "\n".join(lines)
    if len(value) > 1024:
        value = value[:1020] + "..."

    embed.add_field(name=title, value=value, inline=False)
